File: medium/three-sum.ts
// difficulty: medium
// https://leetcode.com/problems/3sum/

function threeSum(nums: number[]): number[][] {
	// Sort
	nums.sort((a, b) => a - b);

	let i = 0;

	const buff: number[][] = [];


	const TARGET_SUM = 0;

	while (i < nums.length - 2) {
		let first = nums[i];

		let j = i + 1;
		let k = nums.length - 1;

		while (j < k) {
			const second = nums[j];
			const third = nums[k];

			let runningSum = first + second + third;

			console.log({ first, second, third, runningSum })

			if (runningSum > TARGET_SUM) {
				k -= 1;
			} else if (runningSum < TARGET_SUM) {
				j += 1;
			} else {
				console.log(first, second, third)
				buff.push([first, second, third]);
				j += 1;
				k -= 1;
			}
		}

		i += 1;
	}

	return buff;
}

console.log(threeSum([-1, 0, 1, 2, -1, -4]))

File: medium/three-sum.py
# difficulty: medium
# https://leetcode.com/problems/3sum/

from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort() # Sort first
        buff = set() # Using a buffer set here to prevent duplicates
        
        i = 0

        while i < len(nums) - 2:
            j = i + 1
            k = len(nums) - 1

            first_num = nums[i]

            while j < k:
                second_num = nums[j]
                third_num = nums[k]

                running_sum = first_num + second_num + third_num

                if running_sum < 0:
                    j += 1
                elif running_sum > 0:
                    k -= 1
                else:
                    buff.add((first_num, second_num, third_num))
                    j += 1
                    k -= 1


            i += 1

        return [list(item) for item in buff]

solution = Solution()
sample = [-1,0,1,2,-1,-4]
ans = solution.threeSum(nums=sample)
print(ans)
